### PR TITLE
templates: use Afterburn for setting GCP hostnames

### PR DESCRIPTION
**Pull request by darkmuggle**
The only platform where over-log hostnames have been encountered is on
GCP. The code has proven buggy, racy, and caused a bunch of BZ.

To unwind this mess, on GCP, the new behavior:
- leaves disabling NetworkManager on GCP
- on each boot run Afterburn to fetch the metadata name
- uses truncation code for ensuring the name is valid

Finally, this DROPS the NetworkManager dispatcher. FCOS/RHCOS is
pursuing a more permanent one.

### CODE REVIEW

**Comment by darkmuggle** [comment]
/hold

I'm NOT sure I like this much better than the other code....its a hack, albeit a lesser hack. I would need to be convinced by others that this isn't horrible.

**Comment by darkmuggle** [comment]
With RHCOS, Afterburn is enable for GCP hostnames. This patch would only applicable for upgrading clusters and would be a bridge between the NetworkManager fix to work its way through.

**Comment by LorbusChris** [comment]
Thank you for putting this up @darkmuggle.

Question: The [node-valid-hostname.service](https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/node-valid-hostname.service.yaml) that calls [set-valid-hostname.sh](https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/files/usr-local-sbin-set-valid-hostname.yaml) still exists in this repo, even after this PR. Is the goal to get rid of that too eventually and handle it elsewhere (in NM/Afterburn)?

**Comment by darkmuggle** [comment]
> Question: The [node-valid-hostname.service](https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/node-valid-hostname.service.yaml) that calls [set-valid-hostname.sh](https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/files/usr-local-sbin-set-valid-hostname.yaml) still exists in this repo, even after this PR. Is the goal to get rid of that too eventually and handle it elsewhere (in NM/Afterburn)?

With this change, the code shifts from checking magic strings to ensuring each GCP node gets a hostname. `valid-hostnames` is needed to wait for the hostname to be set. The _reason_ both this PR and the original code is in the MCO is under the argument that a Linux valid hostname is not a valid cluster-name (i.e 64 character hostnames are valid for Linux, but not the cluster or that <default name> is valid for Linux not for the cluster.) 

Looking at this, I have half a mind to push this unit into FCOS and thereby RHCOS. When we put this into the MCO, we assumed that there was no place where we could fix the hostnames in the OS. We would still need to leave the `valid-hostname` code in the MCO since we're blocking CRI-O and the Kublet until a non-local name is found.

The argument for NOT including this in FCOS/RHCOS is that the OS does not require a hostname, while the cluster does.

**Comment by darkmuggle** [comment]
Logs on a GCP cluster look good:
```
Nov 12 20:25:42 localhost systemd[1]: Started Network Manager Wait Online.
Nov 12 20:25:42 localhost systemd[1]: Starting Set GCP Transient Hostname...
Nov 12 20:25:42 localhost afterburn[1702]: Nov 12 20:25:42.497 INFO Fetching http://metadata.google.internal/computeMetadata/v1/instance/hostname: Attempt #1
Nov 12 20:25:42 localhost afterburn[1702]: Nov 12 20:25:42.502 INFO Fetch successful
Nov 12 20:25:42 localhost systemd[1]: Starting Generate console-login-helper-messages issue snippet...
Nov 12 20:25:42 localhost systemd[1]: Started Generate console-login-helper-messages issue snippet.
Nov 12 20:25:42 localhost bash[1745]: ci-ln-sn85n7t-f76d1-mcv2k-master-2.c.openshift-gce-devel-ci.internal is longer than 63 characters, using trunacated hostname
Nov 12 20:25:42 localhost bash[1749]: setting transient hostname to ci-ln-sn85n7t-f76d1-mcv2k-master-2
Nov 12 20:25:42 localhost systemd[1]: console-login-helper-messages-issuegen.service: Consumed 20ms CPU time
Nov 12 20:25:42 ci-ln-sn85n7t-f76d1-mcv2k-master-2 systemd-hostnamed[1687]: Changed host name to 'ci-ln-sn85n7t-f76d1-mcv2k-master-2'
Nov 12 20:25:46 ci-ln-sn85n7t-f76d1-mcv2k-master-2 chronyd[1440]: Selected source 169.254.169.254
Nov 12 20:25:46 ci-ln-sn85n7t-f76d1-mcv2k-master-2 chronyd[1440]: System clock TAI offset set to 37 seconds
Nov 12 20:25:52 ci-ln-sn85n7t-f76d1-mcv2k-master-2 systemd[1]: NetworkManager-dispatcher.service: Consumed 100ms CPU time
Nov 12 20:26:07 ci-ln-sn85n7t-f76d1-mcv2k-master-2 bash[1729]: Could not set property: Connection timed out
Nov 12 20:26:07 ci-ln-sn85n7t-f76d1-mcv2k-master-2 systemd[1]: Started Set GCP Transient Hostname.
Nov 12 20:26:07 ci-ln-sn85n7t-f76d1-mcv2k-master-2 systemd[1]: Starting Ensure the node hostname is valid for the cluster...
Nov 12 20:26:07 ci-ln-sn85n7t-f76d1-mcv2k-master-2 bash[1756]: waiting for non-localhost hostname to be assigned
Nov 12 20:26:07 ci-ln-sn85n7t-f76d1-mcv2k-master-2 bash[1761]: node identified as ci-ln-sn85n7t-f76d1-mcv2k-master-2
```

**Comment by vrutkovs** [comment]
`okd-e2e-gcp-op` reached install phase, `okd-e2e-upgrade` failed to install initial cluster (unsurprisingly, shouldn't have started that).

:+1: from my side

**Comment by darkmuggle** [comment]
To unblock OKD, this "hack" is better, and @lucab indicated in Slack that this acceptable. Afterburn will write an ephemeral hosts name and no files are modified so removal is straightforward. We get out of checking magic strings, and ensure that GCP has a good hostname. 

/unhold

**Review by yuqi-zhang — APPROVED** [review]
Given the above context I am tentatively in favour of this approach. Should do some upgrade tests etc. on GCP as well

**Comment by sinnykumari** [comment]
/approve
based on offline discussion and benefit this PR brings

File: templates/common/gcp/units/gcp-hostname.service.yaml
Comment (vrutkovs) [inline templates/common/gcp/units/gcp-hostname.service.yaml:5]:
```suggestion
  Description=Set GCP Transient Hostname
  # Removal of this file signals firstboot completion
  ConditionPathExists=!/etc/ignition-machine-config-encapsulated.json
```
nitpick: service should run only after the node has pivoted

**Review by ashcrow — APPROVED** [review]

**Comment by LorbusChris** [comment]
/retest

@darkmuggle just to double check, is it safe to backport this to release-4.6

**Comment by LorbusChris** [comment]
hmm `e2e-gcp-op` timed out.
/test e2e-gcp-op

**Comment by LorbusChris** [comment]
`e2e-gcp-op` timed out again, maybe it's time to bump the timeout a little again?
/retest

**Comment by LorbusChris** [comment]
cluster failed to be provisioned
/retest

**Comment by darkmuggle** [comment]
/retest

The failures in CI are all unrelated to the change....sigh.

**Pull request merged**
